File: apps/ios/Wordocious/Sources/MedalService.swift
import Foundation
import Supabase

/// Client-side daily reward awarding — 1:1 port of the web's
/// checkAndAwardStreakMedals / checkAndAwardPerfectMedal /
/// awardDailyBonusesIfComplete (apps/web/lib/daily-service.ts). The daily
/// TOP-3 podium medals are awarded by a separate server cron
/// (apps/web/app/api/cron/daily-medals), so they already apply to native; only
/// these three were web-client-only and missing on native.
enum MedalService {
    /// The number of solo daily modes a full sweep requires. Keep in sync with
    /// web daily-service.ts DAILY_MODE_COUNT and Android MedalService.kt — a
    /// hardcoded 9 here silently broke sweep detection when the mode count
    /// changed on only two of the three platforms.
    static let dailyModeCount = 9

    private struct MedalInsert: Encodable {
        let user_id, day, game_mode, play_type, medal_type: String
        let composite_score: Int
    }
    private struct IdRow: Decodable { let id: String }

    /// Streak-milestone medals (7 / 30 / 100 day daily-login streak), once ever.
    static func awardStreakMedals(_ client: SupabaseClient, userId: String, day: String) async {
        struct P: Decodable { let daily_login_streak: Int? }
        guard let rows: [P] = try? await client.from("profiles")
            .select("daily_login_streak").eq("id", value: userId).limit(1).execute().value,
            let streak = rows.first?.daily_login_streak else { return }

        let milestones: [(days: Int, type: String)] = [(7, "streak_7"), (30, "streak_30"), (100, "streak_100")]
        for m in milestones where streak >= m.days {
            let existing: [IdRow] = (try? await client.from("medals")
                .select("id").eq("user_id", value: userId).eq("medal_type", value: m.type)
                .limit(1).execute().value) ?? []
            if existing.isEmpty {
                _ = try? await client.from("medals").insert(MedalInsert(
                    user_id: userId, day: day, game_mode: "ALL", play_type: "solo",
                    medal_type: m.type, composite_score: streak)).execute()
            }
        }
    }

    /// Perfect-game medal (per-mode minimum) — once per day+mode.
    static func awardPerfectMedal(_ client: SupabaseClient, userId: String, gameMode: String,
                                  day: String, guessCount: Int, boardsSolved: Int,
                                  totalBoards: Int, completed: Bool) async {
        guard completed else { return }
        let perfect: Bool
        switch gameMode {
        case "DUEL", "PROPERNOUNDLE", "DUEL_6", "DUEL_7": perfect = guessCount == 1
        case "QUORDLE":  perfect = boardsSolved == 4 && guessCount <= 4
        case "OCTORDLE": perfect = boardsSolved == 8 && guessCount <= 8
        case "SEQUENCE": perfect = boardsSolved == 4 && guessCount <= 4
        case "RESCUE":   perfect = boardsSolved == 4 && guessCount <= 4
        case "GAUNTLET": perfect = boardsSolved == 21
        default: perfect = false
        }
        guard perfect else { return }
        let existing: [IdRow] = (try? await client.from("medals")
            .select("id").eq("user_id", value: userId).eq("day", value: day)
            .eq("game_mode", value: gameMode).eq("medal_type", value: "perfect")
            .limit(1).execute().value) ?? []
        guard existing.isEmpty else { return }
        _ = try? await client.from("medals").insert(MedalInsert(
            user_id: userId, day: day, game_mode: gameMode, play_type: "solo",
            medal_type: "perfect", composite_score: guessCount)).execute()
    }

    /// Daily Sweep (+200 XP) / Flawless (+400 XP) bonuses, awarded once when all
    /// 9 daily solo results exist. Adds the XP to the profile and returns the
    /// (sweep, flawless) split so the XP toast can render the distinct
    /// "+200 sweep" / "+400 flawless" chips (web xp-toast.tsx parity).
    static func awardDailyBonusesIfComplete(_ client: SupabaseClient, userId: String) async -> (sweep: Int, flawless: Int) {
        // Delegated to wordocious.com/api/daily/award-bonuses. The app used to
        // write the daily_bonuses row itself, but that row IS the all-time
        // sweep leaderboard (alltime_sweep_leaderboard counts rows with
        // sweep_awarded = true), so a client able to write the flag could award
        // itself sweeps for days it never played. The server recomputes the
        // award from daily_results and grants the XP.
        let day = LeaderboardService.todayLocal()
        guard let token = try? await client.auth.session.accessToken,
              let url = URL(string: "https://wordocious.com/api/daily/award-bonuses")
        else { return (0, 0) }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.httpBody = try? JSONSerialization.data(withJSONObject: ["day": day])

        struct AwardResponse: Decodable {
            let awarded: Bool?
            let sweepAwarded: Bool?
            let flawlessAwarded: Bool?
        }
        guard let (data, _) = try? await URLSession.shared.data(for: req),
              let res = try? JSONDecoder().decode(AwardResponse.self, from: data),
              res.awarded == true
        else { return (0, 0) }

        // XP values stay client-side ONLY to size the toast chips; the profile
        // XP itself was already written by the server.
        return (res.sweepAwarded == true ? 200 : 0,
                res.flawlessAwarded == true ? 400 : 0)
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/MedalService.kt
package com.wordocious.app.data

import com.wordocious.app.todayLocalDate
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.auth.auth
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Client-side medal + daily-bonus awarding — Android port of the web
 * lib/daily-service.ts trio (checkAndAwardStreakMedals, checkAndAwardPerfectMedal,
 * awardDailyBonusesIfComplete) and iOS MedalService.swift. Daily podium medals
 * (gold/silver/bronze) stay server-cron — never awarded client-side.
 *
 * All writes are idempotent: streak medals are once-ever per milestone, perfect
 * medals once per day+mode, and the sweep/flawless bonuses upsert a one-shot
 * `daily_bonuses` row keyed (user_id, day).
 */
object MedalService {
    private val client get() = SupabaseConfig.client

    private const val DAILY_MODE_COUNT = 9
    private const val DAILY_SWEEP_XP = 200
    private const val FLAWLESS_EXTRA_XP = 400

    @Serializable
    private data class MedalInsert(
        @SerialName("user_id") val userId: String,
        val day: String,
        @SerialName("game_mode") val gameMode: String,
        @SerialName("play_type") val playType: String = "solo",
        @SerialName("medal_type") val medalType: String,
        @SerialName("composite_score") val compositeScore: Int,
    )

    @Serializable
    private data class IdRow(val id: String)

    @Serializable
    private data class StreakRow(@SerialName("daily_login_streak") val streak: Int = 0)

    // ── Streak medals (7/30/100-day daily-login streak, once ever) ───────────────
    suspend fun awardStreakMedals(userId: String, day: String) {
        runCatching {
            val streak = client.postgrest["profiles"]
                .select(Columns.raw("daily_login_streak")) { filter { eq("id", userId) }; limit(1) }
                .decodeSingleOrNull<StreakRow>()?.streak ?: return@runCatching
            for ((days, medalType) in listOf(7 to "streak_7", 30 to "streak_30", 100 to "streak_100")) {
                if (streak < days) continue
                val existing = client.postgrest["medals"]
                    .select(Columns.raw("id")) {
                        filter { eq("user_id", userId); eq("medal_type", medalType) }; limit(1)
                    }
                    .decodeList<IdRow>()
                if (existing.isEmpty()) {
                    client.postgrest["medals"].insert(
                        MedalInsert(userId = userId, day = day, gameMode = "ALL", medalType = medalType, compositeScore = streak)
                    )
                }
            }
        }
    }

    // ── Perfect-game medal (min-guess solve, once per day+mode) ──────────────────
    suspend fun awardPerfectMedal(
        userId: String, gameMode: String, day: String,
        guessCount: Int, boardsSolved: Int, totalBoards: Int, completed: Boolean,
    ) {
        if (!completed) return
        val perfect = when (gameMode) {
            "DUEL", "PROPERNOUNDLE", "DUEL_6", "DUEL_7" -> guessCount == 1
            "QUORDLE" -> boardsSolved == 4 && guessCount <= 4
            "OCTORDLE" -> boardsSolved == 8 && guessCount <= 8
            "SEQUENCE" -> boardsSolved == 4 && guessCount <= 4
            "RESCUE" -> boardsSolved == 4 && guessCount <= 4
            "GAUNTLET" -> boardsSolved == 21
            else -> false
        }
        if (!perfect) return
        runCatching {
            val existing = client.postgrest["medals"]
                .select(Columns.raw("id")) {
                    filter {
                        eq("user_id", userId); eq("day", day)
                        eq("game_mode", gameMode); eq("medal_type", "perfect")
                    }
                    limit(1)
                }
                .decodeList<IdRow>()
            if (existing.isEmpty()) {
                client.postgrest["medals"].insert(
                    MedalInsert(userId = userId, day = day, gameMode = gameMode, medalType = "perfect", compositeScore = guessCount)
                )
            }
        }
    }

    // ── Daily Sweep (+200) / Flawless Victory (+400) ─────────────────────────────
    @Serializable
    private data class BonusRow(
        @SerialName("sweep_awarded") val sweepAwarded: Boolean = false,
        @SerialName("flawless_awarded") val flawlessAwarded: Boolean = false,
    )

    @Serializable
    private data class BonusUpsert(
        @SerialName("user_id") val userId: String,
        val day: String,
        @SerialName("sweep_awarded") val sweepAwarded: Boolean,
        @SerialName("flawless_awarded") val flawlessAwarded: Boolean,
        @SerialName("updated_at") val updatedAt: String,
    )

    @Serializable
    private data class CompletedRow(val completed: Boolean)

    @Serializable
    private data class XpRow(val xp: Int = 0, val level: Int = 1)

    /**
     * Award the one-shot daily bonuses if all 9 daily modes are now done.
     * Returns the NEWLY-awarded XP as (sweepXp, flawlessXp) — (200, 400),
     * (200, 0), (0, 400) or (0, 0) — so the XP toast can show distinct chips.
     */
    suspend fun awardDailyBonusesIfComplete(userId: String): Pair<Int, Int> = runCatching {
        // Delegated to wordocious.com/api/daily/award-bonuses. The app used to
        // write the daily_bonuses row itself, but that row IS the all-time
        // sweep leaderboard (alltime_sweep_leaderboard counts rows with
        // sweep_awarded = true), so a client able to write the flag could award
        // itself sweeps for days it never played. The server recomputes the
        // award from daily_results and grants the XP.
        val day = todayLocalDate()
        val token = SupabaseConfig.client.auth.currentSessionOrNull()?.accessToken
            ?: return@runCatching 0 to 0

        val body = withContext(Dispatchers.IO) {
            val conn = (java.net.URL("https://wordocious.com/api/daily/award-bonuses")
                .openConnection() as java.net.HttpURLConnection).apply {
                requestMethod = "POST"
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Authorization", "Bearer $token")
                doOutput = true
                connectTimeout = 10_000
                readTimeout = 10_000
            }
            conn.outputStream.use { it.write("""{"day":"$day"}""".toByteArray()) }
            if (conn.responseCode !in 200..299) return@withContext null
            conn.inputStream.bufferedReader().use { it.readText() }
        } ?: return@runCatching 0 to 0

        val json = kotlinx.serialization.json.Json { ignoreUnknownKeys = true }
            .parseToJsonElement(body).jsonObject
        fun flag(k: String) = json[k]?.jsonPrimitive?.booleanOrNull ?: false
        if (!flag("awarded")) return@runCatching 0 to 0

        // XP values are only used to size the toast chips; the profile XP was
        // already written by the server.
        (if (flag("sweepAwarded")) DAILY_SWEEP_XP else 0) to
            (if (flag("flawlessAwarded")) FLAWLESS_EXTRA_XP else 0)
    }.getOrDefault(0 to 0)
}
